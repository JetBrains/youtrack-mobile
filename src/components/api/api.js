/* @flow */
import qs from 'qs';
import issueFields from './api__issue-fields';
import agileFields from './api__agile-fields';
import Auth from '../auth/auth';
import log from '../log/log';
import ApiHelper from './api__helper';
import {handleRelativeUrl} from '../config/config';
import type {SprintFull, AgileUserProfile, AgileBoardRow, BoardOnList} from '../../flow/Agile';
import type {AppConfigFilled, EndUserAgreement} from '../../flow/AppConfig';
import type {IssueOnList, IssueFull, TransformedSuggestion, SavedQuery, CommandSuggestionResponse} from '../../flow/Issue';
import type {IssueProject, FieldValue, IssueUser, IssueComment} from '../../flow/CustomFields';

const STATUS_UNAUTHORIZED = 401;
const STATUS_OK_IF_MORE_THAN = 200;
const STATUS_BAD_IF_MORE_THATN = 300;

const MAX_QUERY_LENGTH = 2048;

type RequestOptions = {
  parseJson: boolean
};

const defaultRequestOptions: RequestOptions = {
  parseJson: true
};

class Api {
  auth: Auth;
  config: AppConfigFilled;
  youTrackUrl: string;
  youTrackIssueUrl: string;
  youTrackProjectUrl: string;
  youtTrackFieldBundleUrl: string;

  constructor(auth: Auth) {
    this.auth = auth;
    this.config = auth.config;

    this.youTrackUrl = this.config.backendUrl;
    this.youTrackIssueUrl = `${this.youTrackUrl}/api/issues`;
    this.youTrackProjectUrl =`${this.youTrackUrl}/api/admin/projects`;

    this.youtTrackFieldBundleUrl = `${this.youTrackUrl}/api/admin/customFieldSettings/bundles`;
  }

  async makeAuthorizedRequest(url: string, method: ?string, body: ?Object, options: RequestOptions = defaultRequestOptions) {
    assertLongQuery(url);

    const sendRequest = async () => {
      return await fetch(url, {
        method,
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json, text/plain, */*',
          ...this.auth.getAuthorizationHeaders()
        },
        body: JSON.stringify(body)
      });
    };

    let res = await sendRequest();

    if (res.status === STATUS_UNAUTHORIZED) {
      log.info('Looks like the token is expired, will try to refresh', res);
      await this.auth.refreshToken();
      res = await sendRequest();
    }

    if (res.status < STATUS_OK_IF_MORE_THAN || res.status >= STATUS_BAD_IF_MORE_THATN) {
      throw res;
    }

    if (!options.parseJson) {
      return res;
    }
    return await res.json();
  }

  async getIssue(id: string): Promise<IssueFull> {
    const queryString = qs.stringify({
      fields: issueFields.singleIssue.toString()
    }, {encode: false});

    const issue = await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}/${id}?${queryString}`);
    issue.attachments = ApiHelper.convertRelativeUrls(issue.attachments, 'url', this.config.backendUrl);

    return issue;
  }

  async getIssues(query: string = '', $top: number, $skip: number = 0): Promise<IssueOnList> {
    const queryString = qs.stringify({
      query, $top, $skip,
      fields: issueFields.issuesOnList.toString()
    });

    return await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}?${queryString}`);
  }

  async getIssueComments(issueId: string): Promise<Array<IssueComment>> {
    const queryString = qs.stringify({
      fields: issueFields.issueComment.toString()
    });

    const comments = await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}/${issueId}/comments?${queryString}`);

    comments.forEach(comment => {
      comment.author.avatarUrl = handleRelativeUrl(comment.author.avatarUrl, this.config.backendUrl);
    });

    return comments;
  }

  async getIssuesCount(query: string = ''): Promise<number> {
    const queryString = qs.stringify({sync: false, filter: query});
    const countRes = await this.makeAuthorizedRequest(`${this.youTrackUrl}/rest/issue/count?${queryString}`);
    // If server returns -1 it means that counter is not ready yet, should reload
    if (countRes.value === -1) {
      return new Promise(resolve => setTimeout(resolve, 500))
        .then(() => this.getIssuesCount(query));
    }
    return countRes.value;
  }

  async getUserAgreement(): Promise<?EndUserAgreement> {
    const queryString = qs.stringify({fields: 'endUserAgreement(enabled,text,majorVersion,minorVersion)'});
    const res = await this.makeAuthorizedRequest(
      `${this.auth.config.auth.serverUri}/api/rest/settings/public?${queryString}`,
      'GET'
    );

    return res.endUserAgreement;
  }

  async getUserAgreementState(): Promise<Object> {
    const queryString = qs.stringify({fields: issueFields.userConsent});
    return await this.makeAuthorizedRequest(
      `${this.auth.config.auth.serverUri}/api/rest/users/me?${queryString}`,
      'GET',
    );
  }

  async acceptUserAgreement(): Promise<Object> {
    const body = {fields: issueFields.userConsent};
    return await this.makeAuthorizedRequest(
      `${this.auth.config.auth.serverUri}/api/rest/users/endUserAgreementConsent`,
      'POST',
      {
        body
      }
    );
  }

  async getSavedQueries(): Promise<Array<SavedQuery>> {
    const queryString = qs.stringify({fields: issueFields.issueFolder.toString()});
    return await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/savedQueries?${queryString}`);
  }

  async createIssue(issueDraft: IssueOnList) {
    const queryString = qs.stringify({
      draftId: issueDraft.id,
      fields: issueFields.issuesOnList.toString()
    });
    return await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}?${queryString}`, 'POST', {});
  }

  async loadIssueDraft(draftId: string): IssueFull {
    const queryString = qs.stringify({fields: issueFields.singleIssue.toString()});
    const issue = await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/admin/users/me/drafts/${draftId}?${queryString}`);
    issue.attachments = ApiHelper.convertRelativeUrls(issue.attachments, 'url', this.config.backendUrl);
    return issue;
  }

  /**
   * Creates (if issue has no id) or updates issue draft
   * @param issue
   * @returns {Promise}
     */
  async updateIssueDraft(issue: IssueFull): IssueFull {
    const queryString = qs.stringify({fields: issueFields.singleIssue.toString()});

    const updatedIssue = await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/admin/users/me/drafts/${issue.id || ''}?${queryString}`, 'POST', issue);
    updatedIssue.attachments = ApiHelper.convertRelativeUrls(issue.attachments, 'url', this.config.backendUrl);
    return updatedIssue;
  }

  async updateIssueDraftFieldValue(issueId: string, fieldId: string, value: FieldValue) {
    const queryString = qs.stringify({fields: 'id,ringId,value'});
    const body = {id: fieldId, value};
    return await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/admin/users/me/drafts/${issueId}/fields/${fieldId}?${queryString}`, 'POST', body);
  }

  async submitComment(issueId: string, comment: IssueComment) {
    const queryString = qs.stringify({fields: issueFields.issueComment.toString()});
    const url = `${this.youTrackIssueUrl}/${issueId}/comments/${comment.id || ''}?${queryString}`;

    const submittedComment = await this.makeAuthorizedRequest(url, 'POST', comment);
    if (submittedComment.author && submittedComment.author.avatarUrl) {
      submittedComment.author.avatarUrl = handleRelativeUrl(submittedComment.author.avatarUrl, this.config.backendUrl);
    }

    return submittedComment;
  }

  async updateCommentDeleted(issueId: string, commentId: string, deleted: boolean) {
    const queryString = qs.stringify({fields: issueFields.issueComment.toString()});
    const url = `${this.youTrackIssueUrl}/${issueId}/comments/${commentId}?${queryString}`;

    const comment = await this.makeAuthorizedRequest(url, 'POST', {deleted});
    comment.author.avatarUrl = handleRelativeUrl(comment.author.avatarUrl, this.config.backendUrl);

    return comment;
  }

  async deleteCommentPermanently(issueId: string, commentId: string) {
    return this.makeAuthorizedRequest(`${this.youTrackIssueUrl}/${issueId}/comments/${commentId}`, 'DELETE', null, {parseJson: false});
  }

  async getUserFromHub(id: string) {
    const queryString = qs.stringify({fields: 'avatar/url'});
    return await this.makeAuthorizedRequest(`${this.config.auth.serverUri}/api/rest/users/${id}?${queryString}`);
  }

  async getProjects(query: string): Promise<Array<IssueProject>> {
    const queryString = qs.stringify({
      fields: issueFields.projectOnList.toString(),
      query: query
    });
    return await this.makeAuthorizedRequest(`${this.youTrackProjectUrl}?${queryString}`);
  }

  async getProject(projectId: string) {
    const queryString = qs.stringify({
      fields: issueFields.project.toString()
    });
    return await this.makeAuthorizedRequest(`${this.youTrackProjectUrl}/${projectId}?${queryString}`);
  }

  async updateProject(issue: IssueOnList, project: IssueProject) {
    const body = {
      id: issue.id,
      project: project
    };
    return await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}/${issue.id}`, 'POST', body);
  }

  async getCustomFieldUserValues(bundleId: string): Promise<Array<IssueUser>> {
    const queryString = qs.stringify({
      banned: false,
      sort: true,
      fields: issueFields.user.toString()
    });

    const values = await this.makeAuthorizedRequest(
      `${this.youtTrackFieldBundleUrl}/user/${bundleId}/aggregatedUsers?${queryString}`
    );

    return ApiHelper.convertRelativeUrls(values, 'avatarUrl', this.config.backendUrl);
  }

  async getCustomFieldValues(bundleId: string, fieldValueType: string): Promise<Array<Object>> {
    if (fieldValueType === 'user') {
      return this.getCustomFieldUserValues(bundleId);
    }
    const queryString = qs.stringify({
      fields: issueFields.bundle.toString()
    });

    const res = await this.makeAuthorizedRequest(
      `${this.youtTrackFieldBundleUrl}/${fieldValueType}/${bundleId}?${queryString}`
    );
    return res.values;
  }

  async getStateMachineEvents(issueId: string, fieldId: string) {
    const url = `${this.youTrackIssueUrl}/${issueId}/fields/${fieldId}/possibleEvents?fields=id,presentation`;
    return await this.makeAuthorizedRequest(url);
  }

  attachFile(issueId: string, fileUri: string, fileName: string): Promise<XMLHttpRequest> {
    const formDataContent = new FormData(); //eslint-disable-line no-undef
    // $FlowFixMe
    formDataContent.append('photo', {uri: fileUri, name: fileName, type: 'image/binary'});

    return new Promise((resolve, reject) => {
      const STATE_DONE = 4;
      const xhr = new XMLHttpRequest(); //eslint-disable-line no-undef
      xhr.open('POST', `${this.youTrackUrl}/rest/issue/${issueId}/attachment`);
      const headers = this.auth.getAuthorizationHeaders();
      xhr.setRequestHeader('Authorization', headers.Authorization);

      xhr.onreadystatechange = function () {
        if (xhr.readyState !== STATE_DONE) {
          return;
        }
        if (xhr.status >= 200 && xhr.status < 400) {
          log.log('attach result', xhr);
          return resolve(xhr);
        }
        return reject(xhr);
      };
      xhr.send(formDataContent);
    });
  }

  async updateIssueSummaryDescription(issue: IssueFull) {
    const queryString = qs.stringify({fields: 'id,value'});
    const body = {summary: issue.summary, description: issue.description};

    return await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}/${issue.id}?${queryString}`, 'POST', body);
  }

  async updateIssueFieldValue(issueId: string, fieldId: string, value: FieldValue) {
    const queryString = qs.stringify({fields: 'id,ringId,value'});
    const body = {id: fieldId, value};
    return await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}/${issueId}/fields/${fieldId}?${queryString}`, 'POST', body);
  }

  async updateIssueFieldEvent(issueId: string, fieldId: string, event: Object) {
    const queryString = qs.stringify({fields: 'id,ringId,value'});
    const body = {id: fieldId, event};
    return await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}/${issueId}/fields/${fieldId}?${queryString}`, 'POST', body);
  }

  async updateIssueStarred(issueId: string, hasStar: boolean) {
    return await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}/${issueId}/watchers`, 'POST', {hasStar});
  }

  async updateIssueVoted(issueId: string, hasVote: boolean) {
    return await this.makeAuthorizedRequest(`${this.youTrackIssueUrl}/${issueId}/voters`, 'POST', {hasVote});
  }

  async getMentionSuggests(issueIds: Array<string>, query: string) {
    const $top = 10;
    const fields = 'issues(id),users(id,login,fullName,avatarUrl)';
    const queryString = qs.stringify({$top, fields, query});
    const body = {issues:  issueIds.map(id => ({id}))};
    const suggestions = await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/mention?${queryString}`, 'POST', body);
    return ApiHelper.patchAllRelativeAvatarUrls(suggestions, this.config.backendUrl);
  }

  async getCommandSuggestions(issueIds: Array<string>, query: string, caret: number): Promise<CommandSuggestionResponse> {
    const queryString = qs.stringify({fields: issueFields.commandSuggestionFields.toString()});

    return await this.makeAuthorizedRequest(
      `${this.youTrackUrl}/api/commands/assist?${queryString}`,
      'POST',
      {
        query,
        caret,
        issues: issueIds.map(id => ({id}))
      }
    );
  }

  async applyCommand(options: {issueIds: Array<string>, comment?: ?string, command: string}): Promise<any> {
    return await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/commands`, 'POST', {
      query: options.command,
      comment: options.comment,
      issues: options.issueIds.map(id => ({id}))
    });
  }

  //TODO: this is old API usage, move to new one
  async getQueryAssistSuggestions(query: string, caret: number): Promise<Array<TransformedSuggestion>> {
    const queryString = qs.stringify({query, caret});
    const result = await this.makeAuthorizedRequest(`${this.youTrackUrl}/rest/search/underlineAndSuggest?${queryString}`);

    return ApiHelper.convertQueryAssistSuggestions(result.suggest.items);
  }

  async getAgileUserProfile(): Promise<AgileUserProfile> {
    const queryString = qs.stringify({
      fields: agileFields.agileUserProfile.toString()
    });
    return await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/agileUserProfile?${queryString}`);
  }

  async getSprint(boardId: string, sprintId: string, top: number = 100, skip: number = 0): Promise<SprintFull> {
    const queryString = qs.stringify({
      fields: agileFields.sprint.toString(),
      $topSwimlanes: top,
      $skipSwimlanes: skip
    }, {encode: false});
    const sprint = await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/agiles/${boardId}/sprints/${sprintId}?${queryString}`);
    return ApiHelper.patchAllRelativeAvatarUrls(sprint, this.config.backendUrl);
  }

  async getSwimlanes(boardId: string, sprintId: string, top: number, skip: number = 0): Promise<Array<AgileBoardRow>> {
    const queryString = qs.stringify({
      fields: `trimmedSwimlanes(${agileFields.row.toString()})`,
      $topSwimlanes: top,
      $skipSwimlanes: skip
    });

    const board = await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/agiles/${boardId}/sprints/${sprintId}/board?${queryString}`);
    const swimlanes = board.trimmedSwimlanes;
    return ApiHelper.patchAllRelativeAvatarUrls(swimlanes, this.config.backendUrl);
  }

  async updateRowCollapsedState(boardId: string, sprintId: string, row: Object): Promise<Object> {
    const isOrphan = row.id === 'orphans';
    const url = isOrphan ?
      `${this.youTrackUrl}/api/agiles/${boardId}/sprints/${sprintId}/board/orphanRow` :
      `${this.youTrackUrl}/api/agiles/${boardId}/sprints/${sprintId}/board/swimlanes/${row.id}`;

    return await this.makeAuthorizedRequest(`${url}`, 'POST', {
      $type: row.$type,
      id: isOrphan ? null : row.id,
      collapsed: row.collapsed
    });
  }

  async updateColumnCollapsedState(boardId: string, sprintId: string, column: Object): Promise<Object> {
    return await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/agiles/${boardId}/sprints/${sprintId}/board/columns/${column.id}`,
    'POST',
    {
      collapsed: column.collapsed
    });
  }

  async getSprintList(boardId: string): Promise<Object> {
    const queryString = qs.stringify({
      fields: agileFields.sprintShort.toString()
    });
    return await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/agiles/${boardId}/sprints?${queryString}`);
  }

  async getAgileBoardsList(): Promise<Array<BoardOnList>> {
    const queryString = qs.stringify({
      fields: agileFields.boardOnList.toString()
    });
    return await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/agiles?${queryString}`);
  }

  async saveLastVisitedSprint(sprintId: string): Promise<Object> {
    return await this.makeAuthorizedRequest(`${this.youTrackUrl}/api/agileUserProfile`, 'POST', {
      visitedSprints: [{id: sprintId}]
    });
  }

  async getIssueDraftForAgileCell(boardId: string, sprintId: string, columnId: string, cellId: string): Promise<{id: string}> {
    const queryString = qs.stringify({fields: 'id'});
    const url =`${this.youTrackUrl}/api/agiles/${boardId}/sprints/${sprintId}/board/columns/${columnId}/cells/${cellId}/draftIssue?${queryString}`;
    return await this.makeAuthorizedRequest(url, 'POST', {});
  }

  async getVisibilityOptions(issueId: string): Promise<any> {
    const queryString = qs.stringify({
      $top: 50,
      fields: issueFields.getVisibility.toString()
    });
    const url =`${this.youTrackUrl}/api/visibilityGroups?${queryString}`;
    const visibilityOptions = await this.makeAuthorizedRequest(url, 'POST', {issues: [{id: issueId}]});
    visibilityOptions.visibilityUsers = ApiHelper.convertRelativeUrls((visibilityOptions.visibilityUsers || []), 'avatarUrl', this.config.backendUrl);
    return visibilityOptions;
  }

  async getNotificationsToken(): Promise<string> {
    const url = `${this.youTrackUrl}/api/userNotifications/subscribe?fields=token`;
    const res = await this.makeAuthorizedRequest(url, 'POST');
    return res.token;
  }
}

/**
 * https://youtrack.jetbrains.com/issue/YTM-261
 * http://www.mytecbits.com/microsoft/iis/iis-changing-maxquerystring-and-maxurl
 */
function assertLongQuery(url: string) {
  const [, ...queryParts] = url.split('?');
  const query = queryParts.join('');
  if (query.length > MAX_QUERY_LENGTH) {
    log.warn(`Query length (${query.length}) is longer than ${MAX_QUERY_LENGTH}. This doesn't work on some servers`, url);
  }
}

export default Api;
