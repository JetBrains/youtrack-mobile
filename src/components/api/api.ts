import * as qs from 'qs';
import AgileAPI from './api__agile';
import ApiHelper from './api__helper';
import ArticlesAPI from './api__articles';
import BaseAPI from './api__base';
import CustomFieldsAPI from './api__custom-fields';
import FilterFields from 'components/api/api__filter-fields';
import InboxAPI from './api__inbox';
import IssueAPI from './api__issue';
import issueFields from './api__issue-fields';
import IssueFolderAPI from './api__issue-folder';
import IssuesAPI from './api__issues';
import MentionsAPI from './api__mention';
import ProjectsAPI from './api__projects';
import SavedQueries from 'components/api/api__saved-queries';
import SearchAPI from './api__search';
import UserAPI from './api__user';
import UserGroupAPI from './api__user-group';

import type Auth from '../auth/oauth2';
import type {EndUserAgreement} from 'types/AppConfig';
import type {IssueProject} from 'types/CustomFields';
import type {CommandSuggestionResponse} from 'types/Issue';
import type {User} from 'types/User';


class API extends BaseAPI {
  youTrackProjectUrl: string;
  youtTrackFieldBundleUrl: string;
  agile: AgileAPI;
  articles: ArticlesAPI;
  customFields: CustomFieldsAPI;
  filterFields: FilterFields;
  inbox: InboxAPI;
  issue: IssueAPI;
  issueFolder: IssueFolderAPI;
  issues: IssuesAPI;
  mentions: MentionsAPI;
  projects: ProjectsAPI;
  savedQueries: SavedQueries;
  search: SearchAPI;
  user: UserAPI;
  userGroup: UserGroupAPI;

  constructor(auth: Auth) {
    super(auth);
    this.agile = new AgileAPI(auth);
    this.articles = new ArticlesAPI(auth);
    this.customFields = new CustomFieldsAPI(auth);
    this.filterFields = new FilterFields(auth);
    this.inbox = new InboxAPI(auth);
    this.issue = new IssueAPI(auth);
    this.issueFolder = new IssueFolderAPI(auth);
    this.issues = new IssuesAPI(auth);
    this.mentions = new MentionsAPI(auth);
    this.projects = new ProjectsAPI(auth);
    this.savedQueries = new SavedQueries(auth);
    this.search = new SearchAPI(auth);
    this.user = new UserAPI(auth);
    this.userGroup = new UserGroupAPI(auth);
    this.youTrackProjectUrl = `${this.youTrackUrl}/api/admin/projects`;
    this.youtTrackFieldBundleUrl = `${this.youTrackUrl}/api/admin/customFieldSettings/bundles`;
  }

  async getUserAgreement(): Promise<EndUserAgreement | null | undefined> {
    const queryString = qs.stringify({
      fields: 'endUserAgreement(enabled,text,majorVersion,minorVersion)',
    });
    const res = await this.makeAuthorizedRequest(
      `${this.auth.config.auth.serverUri}/api/rest/settings/public?${queryString}`,
      'GET',
    );
    return res.endUserAgreement;
  }

  async acceptUserAgreement(): Promise<Record<string, any>> {
    const body = {
      fields: issueFields.userConsent,
    };
    return await this.makeAuthorizedRequest(
      `${this.auth.config.auth.serverUri}/api/rest/users/endUserAgreementConsent`,
      'POST',
      {
        body,
      },
    );
  }

  async getProjects(query: string): Promise<Array<IssueProject>> {
    const queryString = qs.stringify({
      fields: issueFields.projectOnList.toString(),
      query: query,
    });
    return await this.makeAuthorizedRequest(
      `${this.youTrackProjectUrl}?${queryString}`,
    );
  }

  async getCustomFieldUserValues(bundleId: string): Promise<User[]> {
    const queryString = qs.stringify({
      banned: false,
      sort: true,
      fields: issueFields.user.toString(),
    });
    const values: User[] = await this.makeAuthorizedRequest(
      `${this.youtTrackFieldBundleUrl}/user/${bundleId}/aggregatedUsers?${queryString}`,
    );
    return ApiHelper.convertRelativeUrls(
      values,
      'avatarUrl',
      this.config.backendUrl,
    ) as User[];
  }

  async getCustomFieldValues(
    bundleId: string,
    fieldValueType: string,
  ): Promise<Array<Record<string, any>>> {
    if (fieldValueType === 'group') {
      return this.userGroup.getAllUserGroups();
    }

    if (fieldValueType === 'user') {
      return this.getCustomFieldUserValues(bundleId);
    }

    const queryString = API.createFieldsQuery(issueFields.bundleValues, {
      $includeArchived: false,
      sort: true,
    });
    return await this.makeAuthorizedRequest(
      `${this.youtTrackFieldBundleUrl}/${fieldValueType}/${bundleId}/values?${queryString}`,
    );
  }

  async getStateMachineEvents(issueId: string, fieldId: string): Promise<any> {
    const url = `${this.youTrackIssueUrl}/${issueId}/fields/${fieldId}/possibleEvents?fields=id,presentation`;
    return await this.makeAuthorizedRequest(url);
  }

  async getCommandSuggestions(
    issueIds: string[],
    query: string,
    caret: number,
  ): Promise<CommandSuggestionResponse> {
    const queryString = qs.stringify({
      fields: issueFields.commandSuggestionFields.toString(),
    });
    return await this.makeAuthorizedRequest(
      `${this.youTrackUrl}/api/commands/assist?${queryString}`,
      'POST',
      {
        query,
        caret,
        issues: issueIds.map(id => ({
          id,
        })),
      },
    );
  }

  async applyCommand(options: {
    issueIds: string[];
    comment?: string | null | undefined;
    command: string;
  }): Promise<any> {
    return await this.makeAuthorizedRequest(
      `${this.youTrackUrl}/api/commands`,
      'POST',
      {
        query: options.command,
        comment: options.comment,
        issues: options.issueIds.map(id => ({
          id,
        })),
      },
    );
  }

  async getNotificationsToken(): Promise<string> {
    const url = `${this.youTrackUrl}/api/userNotifications/subscribe?fields=token`;
    const res = await this.makeAuthorizedRequest(url, 'POST');
    return res.token;
  }

  async subscribeToFCMNotifications(
    konnectorURL: string,
    youtrackToken: string,
    deviceToken: string,
  ): Promise<string> {
    const url = `${konnectorURL}/ring/fcmPushNotifications`;
    return await this.makeAuthorizedRequest(url, 'POST', {
      youtrackToken: youtrackToken,
      deviceToken: deviceToken,
    });
  }

  async unsubscribeFromFCMNotifications(
    konnectorURL: string,
    deviceToken: string,
  ): Promise<any> {
    return this.makeAuthorizedRequest(
      `${konnectorURL}/ring/fcmPushNotifications/unsubscribe`,
      'POST',
      {
        deviceToken: deviceToken,
      },
    );
  }

  async subscribeToIOSNotifications(
    konnectorURL: string,
    youtrackToken: string,
    deviceToken: string,
  ): Promise<string> {
    const url = `${konnectorURL}/ring/pushNotifications`;
    return await this.makeAuthorizedRequest(url, 'POST', {
      token: youtrackToken,
      appleDeviceId: deviceToken,
    });
  }

  async unsubscribeFromIOSNotifications(
    konnectorURL: string,
    deviceToken: string,
  ): Promise<any> {
    return this.makeAuthorizedRequest(
      `${konnectorURL}/ring/pushNotifications/unsubscribe`,
      'POST',
      {
        deviceToken: deviceToken,
      },
    );
  }

  async getWorkTimeSettings(): Promise<Record<string, any>> {
    const fields = 'id,daysAWeek,workDays,minutesADay';
    const url = `${this.youTrackUrl}/api/admin/timeTrackingSettings/workTimeSettings?fields=${fields}`;
    return await this.makeAuthorizedRequest(url, 'GET');
  }
}

export default API;
