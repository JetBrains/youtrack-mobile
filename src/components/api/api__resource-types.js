/* @flow */

export const ResourceTypes = {
  AGILE: 'jetbrains.youtrack.agile.settings.Agile',

  ARTICLE: 'Article',
  ARTICLE_DRAFT: 'ArticleDraft',

  COMMENT_REACTIONS_FEED_ITEM: 'CommentReactionsFeedItem',

  EVENT_GROUP: 'jetbrains.youtrack.event.gaprest.ActivityItemGroup',

  ISSUE: 'jetbrains.charisma.persistent.Issue',
  ISSUE_COMMENT: 'jetbrains.charisma.persistent.IssueComment',
  ISSUE_FOLDER_SAVED_QUERY: 'jetbrains.charisma.persistent.issueFolders.SavedQuery',
  ISSUE_FOLDER_TAG: 'jetbrains.charisma.persistent.issueFolders.IssueTag',

  PROJECT: 'jetbrains.charisma.persistent.Project',

  USER: 'jetbrains.charisma.persistence.user.User',
  USER_APPEARANCE_PROFILE: 'AppearanceUserProfile',
  USER_GENERAL_PROFILE: 'GeneralUserProfile',
  USER_GROUP: 'jetbrains.charisma.persistent.security.UserGroup',

  VISIBILITY_GROUP: 'jetbrains.charisma.persistent.security.VisibilityGroups',
  VISIBILITY_LIMITED: 'jetbrains.charisma.persistent.visibility.LimitedVisibility',
  VISIBILITY_UNLIMITED: 'jetbrains.charisma.persistent.visibility.UnlimitedVisibility',

  WORK_ITEM: 'jetbrains.youtrack.timetracking.gaprest.IssueWorkItem',
  DRAFT_WORK_ITEM: 'DraftWorkItem',

  DRAFT_ISSUE_COMMENT: 'DraftIssueComment',
  DRAFT_ARTICLE_COMMENT: 'DraftArticleComment'
};

type HasMethodName = 'comment' | 'user' | 'userGroup' | 'project' | 'savedSearch' | 'tag' | 'agile';
type Entity = $Shape<{ $type: string }>;

export const hasType: Object = function (type: string) {
  return function (it: Entity): boolean {
    return it ? it.$type === type || it.$type === getShortEntityType(type) : false;
  };
};

hasType.agile = hasType(ResourceTypes.ISSUE_FOLDER_TAG);
hasType.article = hasType(ResourceTypes.ARTICLE);
hasType.articleDraft = hasType(ResourceTypes.ARTICLE_DRAFT);
hasType.comment = hasType(ResourceTypes.ISSUE_COMMENT);
hasType.commentReaction = hasType(ResourceTypes.COMMENT_REACTIONS_FEED_ITEM);
hasType.issue = hasType(ResourceTypes.ISSUE);
hasType.project = hasType(ResourceTypes.PROJECT);
hasType.savedSearch = hasType(ResourceTypes.ISSUE_FOLDER_SAVED_QUERY);
hasType.tag = hasType(ResourceTypes.ISSUE_FOLDER_TAG);
hasType.user = hasType(ResourceTypes.USER);
hasType.userGroup = hasType(ResourceTypes.USER_GROUP);
hasType.visibilityLimited = hasType(ResourceTypes.VISIBILITY_LIMITED);
hasType.work = hasType(ResourceTypes.WORK_ITEM);


export function filterArrayByType(array: Array<Entity>, methodName: HasMethodName): Array<Entity> {
  return (array || []).filter((it: Entity) => hasType[methodName] && hasType[methodName](it));
}


export const addTypes = function (type: string): Array<string> {
  return [].concat(type).concat(getShortEntityType(type));
};

export function getShortEntityType(type: string): string {
  return type.split('.').pop();
}
