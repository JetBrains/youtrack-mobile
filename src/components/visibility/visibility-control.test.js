import React from 'react';

import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

import {ResourceTypes} from '../api/api__resource-types';
import VisibilityControl from './visibility-control';

describe('<VisibiltyControl/>', () => {

  let wrapper;
  let instance;
  let visibilityMock;

  beforeEach(() => {
    visibilityMock = {
      $type: ResourceTypes.VISIBILITY_UNLIMITED
    };

    render();
  });


  describe('Render', () => {
    it('should match a snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should initialize state', () => {
      expect(findByTestId('visibilityControl')).toHaveLength(1);
      expect(findByTestId('visibilityControlButton')).toHaveLength(1);
    });
  });


  describe('Init', () => {
    it('should initialize state', () => {
      expect(instance.state.visibility).toEqual(visibilityMock);
      expect(instance.state.isSelectVisible).toEqual(false);
    });
  });


  describe('Visibility', () => {
    let restrictedVisibilityMock;
    let onApplyMock;
    let onSubmitMock;

    beforeEach(() => {
      restrictedVisibilityMock = {
        $type: ResourceTypes.VISIBILITY_LIMITED,
        permittedGroups: [{}]
      };
      onApplyMock = jest.fn();
      onSubmitMock = jest.fn();

    });

    it('should reset visibility', () => {
      render(onApplyMock, onSubmitMock);

      instance.resetVisibility();

      expect(instance.state.visibility).toEqual(null);
    });

    it('should submit visibility', () => {
      render(onApplyMock, onSubmitMock);

      instance.updateVisibility(restrictedVisibilityMock);

      expect(onSubmitMock).toHaveBeenCalledWith(restrictedVisibilityMock);
      expect(onApplyMock).not.toHaveBeenCalledWith(restrictedVisibilityMock);
    });

    it('should apply visibility', () => {
      render(onApplyMock, null);

      instance.updateVisibility(restrictedVisibilityMock);

      expect(onSubmitMock).not.toHaveBeenCalledWith(restrictedVisibilityMock);
      expect(onApplyMock).toHaveBeenCalledWith(restrictedVisibilityMock);
    });

    it('should update state with a new visibility', () => {
      instance.updateVisibility(restrictedVisibilityMock);

      expect(instance.state.visibility).toEqual(restrictedVisibilityMock);
    });
  });


  function findByTestId(testId) {
    return wrapper && wrapper.find({testID: testId});
  }

  function doShallow(visibility, onApply = () => {}, onSubmit = () => {}) {
    return shallow(
      <VisibilityControl
        issueId={'issueID'}
        visibility={visibility}
        onApply={onApply}
        onSubmit={onSubmit}
      />
    );
  }

  function render(onApply, onSubmit) {
    wrapper = doShallow(visibilityMock, onApply, onSubmit);
    instance = wrapper.instance();
  }
});
