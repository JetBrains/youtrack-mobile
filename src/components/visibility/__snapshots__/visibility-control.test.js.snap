// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VisibiltyControl/> Render should match a snapshot 1`] = `
<View
  testID="visibilityControl"
>
  <View
    hitSlop={
      Object {
        "bottom": 8,
        "left": 8,
        "right": 8,
        "top": 8,
      }
    }
    style={
      Array [
        undefined,
        null,
      ]
    }
    testID="visibilityControlButton"
  >
    <TouchableOpacity
      activeOpacity={0.2}
      hitSlop={
        Object {
          "bottom": 8,
          "left": 8,
          "right": 8,
          "top": 8,
        }
      }
      onPress={[Function]}
    >
      <Text>
        Visible to All Users
      </Text>
      <IconAngleDown
        color="#737577"
        size={20}
      />
    </TouchableOpacity>
  </View>
</View>
`;
